### ISSUE TITLE
pickle.loads() of pickled numpy.array can corrupt memory (migrated from Trac #1708)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3259]
Original ticket http://projects.scipy.org/numpy/ticket/1708
Reported 2011-01-10 by trac user quarl, assigned to unknown.

We discovered that pickle.loads() on a pickled numpy.ndarray can lead to memory corruption.  I can reproduce the problem with arrays of size 1 (length=1, dtype=bool or byte) pickled using protocols 1 or 2:
    pickle.loads(pickle.dumps(numpy.array([True]), protocol=-1))

In my tests, this affects Solaris and Linux, Sparc, x86, and x86_64, Python versions up to 2.6.6, numpy version 1.4.1.

After unpickling the size-1 array into 'data', I replace data[0] with an arbitrary byte (0xbb in the example below) and it overwrites some unrelated data (regexp buffers in the example below).  If the original data contains array([2]) instead of array([1]), then it affects "\x02" in the regexp buffer instead of "\x01".

The second use of pickle.loads() then gets confused.

```
#!/usr/bin/env python

import numpy
import pickle
import re

# Unpickle a bool or byte array of length 1:
data = numpy.array([1], dtype='b') # similar issue with numpy.array([True])
blob = pickle.dumps(data, protocol=1) # protocol 1, -1 exhibit bug, but 0 is 'safe'
data = pickle.loads(blob)

print repr(re.sub("a(.)", "\x01\\1", "a_"))
# '\x01_' [correct]

data[0] = 0xbb

print repr(re.sub("a(.)", "\x01\\1", "a_"))
# '\xbb_' [?!]

lose = pickle.loads(blob) # crashes
```

**Comment by thouis** [comment]
Comment in Trac by trac user quarl, 2011-01-10

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-03-04

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-04-02

Fixed in commit:7b39b9e5

It's not really memory corruption -- the problem was that Numpy happily used also interned strings as buffers when unpickling.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-04-02

Same as #1233
